### DISCUSSION TITLE
삼각형 스트립 vs 인덱스 삼각형 리스트

### DISCUSSION THREAD

**Discussion by Lcat12345** [MyeongWoonJang/cpp-study#99] [Show and tell]
`삼각형 스트립`보다 `인덱스 삼각형 리스트`가 더 효율적이다.
삼각형을 처리할때 정점처리를 더 적게 한다. 
즉, `ACMR`이 `인덱스 삼각형 리스트` 쪽이 낮다.

캐쉬에 공간이 충분한 경우 `인덱스 삼각형 리스트`의 경우 
단순하게 생각하면 **정점 처리를 해야 될때도 있고 안해도 될때도 있다.**
그러나 `삼각형 스트립`은 **무조건 한 정점**은 처리해야 하므로 성능이 떨어진다.

캐쉬에 공간이 없어서 정점을 계속 찾아야 한다면 인덱스 삼각형 리스트의 ACMR이 3까지 올라가므로 이 경우 삼각형 스트립이 더 좋다.

**Comment by Woon-2** [comment]
`indexed triangle list`는 모든 정점을 단 한 번씩만 저장합니다.   
각 정점마다 인덱스를 부여하니 2/4바이트의 메모리 `overhead`가 발생합니다.

`triangle strip`은 모든 정점을 단 한 번씩 저장하는 경우도 가능하지만, 대개 그렇지 않습니다.   
<div align="center"><image src="https://user-images.githubusercontent.com/73771162/201447856-2a89302e-df98-45c6-961e-509ddba92197.png"></div>
<div align="center"> source:

[ResearchGate](https://www.researchgate.net/figure/A-The-triangle-strip-B-the-triangle-fan-and-C-the-generalized-triangle-strip_fig2_222656093) </div>

A의 경우 새로운 정점 하나씩만 추가하여 `strip`을 구성하지만,   
B의 경우 `v1`, `v2`, `v3`로 삼각형을 만들 수 없어 `v0`에 해당하는 정점을 새로 만들 것이며,
C의 경우 `v6`, `v7`, `v8`로 삼각형을 만들 수 없고,   
새로운 삼각형을 `v9`, `v5`, `v3`를 이어서 만든다 할 때 `v3`에 해당하는 정점을 새로 만들게 됩니다.   
두꺼운 기하구조를 만들 수록 `triangle strip`의 단점은 쉽게 드러납니다.   

한 정점은 일반적으로 `float` 8개로 이루어진 32바이트(x,y,z,r,g,b,u,v) 구조체니,    
중복 이용되는 정점 하나당 32바이트의 메모리 `overhead`가 발생하는 겁니다.

그러나 `indexed triangle list`는 모든 정점에 대해 `overhead`가 발생하는 반면,   
`triangle strip`은 몇몇 정점에 대해서만 `overhead`가 발생하죠.   

문제는 실제 메시들을 그릴 때, 몇몇 정점이 아니라 엄청난 수의 정점들이 중복 이용될 수 있는데도 새로 만들어진다는 것입니다.   
삼각형 하나는 매우 작아서, 실제 메시들은 두꺼운 기하구조인 경우가 많거든요.

우리 눈에는 똑같은 정점인데도 다른 정점으로 메모리까지 새로 만들어 처리하니,   
기존의 메모리를 이용하지 않는다 + 새로운 메모리를 이용한다는 점에서 캐시 효율이 굉장히 떨어지죠.   

<br>

`indexed triangle list`는 같은 인덱스를 참조하는 삼각형 A와 B를 한참의 간격을 두고 그린다거나,   
가리키는 정점들의 메모리 상 거리가 매우 먼 인덱스 1, 3000, 6000으로 삼각형을 구성한다거나 하면   
캐시 효율이 떨어지게 됩니다.   

최근의 `indexed triangle list`를 구성하는 알고리즘은 공간 지역성, 시간 지역성을 고려해   
캐시 효율을 최적화합니다.

`indexed triangle list`가 최선의 선택인 경우가 99%라고 해도 무리가 없습니다.

**Reply by Lcat12345** [reply]
> 각 정점마다 인덱스를 부여하니 2/4바이트의 메모리 overhead가 발생합니다.

- 왜 2/4바이트 인가요?

- 캐쉬 미스가 곧 오버헤드라고 생각해도 될까요.

**Reply by Woon-2** [reply]
- 인덱스를 부호 없는 2바이트 정수로 구현할 수도 있고, 부호 없는 4바이트 정수로 구현할 수도 있기 때문입니다.
- 아닙니다.   
  제가 글에서 언급한 `overhead`는 캐시와 관련 없이, 그저 메모리를 얼마만큼 추가로 잡아먹느냐에 대한 `overhead`입니다.

  - `index`는 정점마다 하나씩 부여되므로, 정점이 n개 있다 하면 2n 또는 4n 바이트의 메모리 `overhead`가 발생합니다.
  - `triangle strip`이 기존에 저장한 정점인데도 불구하고 새롭게 정점을 저장하면, 중복 저장 횟수를 m번이라 할때 일반적으로 32m 바이트의 메모리 `overhead`가 발생합니다.

중복 저장이라 함은 이런 겁니다.

```cpp
struct vertex_initializer { ... };
class vertex
{
    explicit vertex(const vertex_initializer& vi) { ... }
    
    ...
};

vertex vertex_array[1024];    // 배열의 크기는 별 의미 없습니다.

void store_vertex(const vertex_initializer& vi)
{
    static int n = 0;
    vertex_array[n++] = vertex{ vi };
}

void draw(int i1, int i2, int i3)
{
    ... // vertex_array[i1], vertex_array[i2], vertex_array[3]로 구성된 삼각형을 그립니다.
}

int main()
{
    vertex_initializer v0 { ... };
    vertex_initializer v1 { ... };
    vertex_initializer v2 { ... };
    vertex_initializer v3 { ... };
    vertex_initializer v4 { ... };

    store_vertex(v0);
    store_vertex(v1);
    store_vertex(v2);

    draw(0, 1, 2);

    store_vertex(v3);

    draw(1, 2, 3);

    store_vertex(v4);

    draw(2, 3, 4);

    store_vertex(v0);    // 같은 값에 대한 중복 저장!
                         // vertex_array[0]와 vertex_array[5]는 값은 같지만, 분명히 서로 다른 메모리를 차지하고 있습니다.
                         // 슬픕니다. 우리야 실제로 이 둘이 같은 걸 알지만
                         // 컴퓨터는 그저 시킨대로 메모리를 새로 잡아먹을 뿐입니다.

    draw(3, 4, 5);    // draw(3, 4, 0)이 아닙니다.
}
```

같은 메모리를 참조해도 될 것을 다른 메모리를 참조함으로써 캐시 효율도 연달아 떨어집니다.   

이미 참조했던 메모리는 캐시에 올라와 있을 건데,   
새 메모리를 참조하고 있잖아요.   

값이 같은 거지, 메모리 주소가 같질 않으니까 바보 같이 캐시에 있는 걸 냅두고 메모리에서 꺼내와버립니다.

캐시 미스 확률이요?   
100%입니다.   

캐시에 없는 메모리를 꺼내오는데 어떻게 캐시가 적중할 수 있을까요.   

그래놓고선 꺼내온 메모리를 좋아라 캐시에 새로 저장할 겁니다.   
"다음에 이용할지도 몰라! 헤헤" 하면서요.   
이미 캐시에 값이 같은 데이터가 있다는 것을 죽을 때까지 모를 겁니다.

그러면 또 연쇄적으로 캐시 효율이 떨어집니다.   
캐시에 새로 저장했다는 것은, 그 부분에 기존에 저장되어 있던 내용이 날아갔다는 겁니다.   

메모리를 많이 차지하는 게 캐시 효율에 미치는 영향이 딱 이런 겁니다.   

시험 범위가 1페이지인 게 외우기 쉽나요,   
시험 범위가 600페이지인 게 외우기 쉽나요?

여러분의 뇌용량이 정해져 있어 기억하는 데 한계가 있는 것과,   
캐시 용량이 정해져 있어 기억하는 데 한계가 있는 것은 정확히 같습니다.   

결국에 캐시도 새로운 정보를 기록하며 기존의 정보를 잊어버릴 것이라구요.

메모리 `overhead`가 캐시 미스를 유발할 가능성이 있는 것이지,
메모리 `overhead` = 캐시 미스라고 생각하시면 안 됩니다.

**Reply by Lcat12345** [reply]
`triangle strip`은 재사용 할수있는, 캐시에 올라간 정점을 이용하지 않고 새로 저장하기 때문에 
또한 그 정점은 무려 32바이트나 차지해서 ` indexed triangle list`보다 훨씬 안좋다는 내용인것 같네요.
